Index: bfd/acinclude.m4
===================================================================
--- bfd/acinclude.m4	(Revision 5)
+++ bfd/acinclude.m4	(Arbeitskopie)
@@ -21,7 +21,7 @@
 [AC_REQUIRE([AC_CANONICAL_TARGET])
 case "${host}" in
 changequote(,)dnl
-*-*-msdos* | *-*-go32* | *-*-mingw32* | *-*-cygwin* | *-*-windows*)
+*-*-msdos* | *-*-go32* | *-*-mingw32* | *-*-cygwin* | *-*-windows* | *-*-mint*)
 changequote([,])dnl
   AC_DEFINE(USE_BINARY_FOPEN, 1, [Use b modifier when opening binary files?]) ;;
 esac])dnl
Index: bfd/config.bfd
===================================================================
--- bfd/config.bfd	(Revision 5)
+++ bfd/config.bfd	(Arbeitskopie)
@@ -920,6 +920,10 @@
     # targ_selvecs=m68kmach3_vec
     # targ_cflags=-DSTAT_FOR_EXEC
     ;;
+  m68*-*-mint*)
+    targ_defvec=m68k_elf32_vec
+    targ_underscore=yes
+    ;;
   m68*-hp*-netbsd*)
     targ_defvec=m68k_aout_4knbsd_vec
     targ_selvecs="m68k_aout_nbsd_vec m68k_aout_hp300bsd_vec sparc_aout_sunos_be_vec"
Index: bfd/configure
===================================================================
--- bfd/configure	(Revision 5)
+++ bfd/configure	(Arbeitskopie)
@@ -12200,7 +12200,7 @@
       *)   PKGVERSION="($withval) " ;;
      esac
 else
-  PKGVERSION="(GNU Binutils) "
+  PKGVERSION="(GNU Binutils for MiNT 20170202) "
 
 fi
 
@@ -13761,7 +13761,7 @@
 
 
 case "${host}" in
-*-*-msdos* | *-*-go32* | *-*-mingw32* | *-*-cygwin* | *-*-windows*)
+*-*-msdos* | *-*-go32* | *-*-mingw32* | *-*-cygwin* | *-*-windows* | *-*-mint*)
 
 $as_echo "#define USE_BINARY_FOPEN 1" >>confdefs.h
  ;;
Index: bfd/elf32-m68k.c
===================================================================
--- bfd/elf32-m68k.c	(Revision 5)
+++ bfd/elf32-m68k.c	(Arbeitskopie)
@@ -1052,7 +1052,7 @@
 				 bfd_boolean linker ATTRIBUTE_UNUSED)
 {
   int mach = bfd_get_mach (abfd);
-  unsigned long e_flags = elf_elfheader (abfd)->e_flags;
+  unsigned long e_flags = elf_elfheader (abfd)->e_flags = EF_M68K_M68000; /* for m68k-atari-mint allways M68000 */
 
   if (!e_flags)
     {
Index: binutils/configure
===================================================================
--- binutils/configure	(Revision 5)
+++ binutils/configure	(Arbeitskopie)
@@ -13599,7 +13599,7 @@
 
 
 case "${host}" in
-*-*-msdos* | *-*-go32* | *-*-mingw32* | *-*-cygwin* | *-*-windows*)
+*-*-msdos* | *-*-go32* | *-*-mingw32* | *-*-cygwin* | *-*-windows* | *-*-mint*)
 
 $as_echo "#define USE_BINARY_FOPEN 1" >>confdefs.h
  ;;
Index: binutils/dlltool.c
===================================================================
--- binutils/dlltool.c	(Revision 5)
+++ binutils/dlltool.c	(Arbeitskopie)
@@ -4442,7 +4442,7 @@
       if (*cp == '-')
 	dash = cp;
       if (
-#if defined(__DJGPP__) || defined (__CYGWIN__) || defined(__WIN32__)
+#if defined(__DJGPP__) || defined (__CYGWIN__) || defined(__WIN32__) || defined(__MINT__)
 	  *cp == ':' || *cp == '\\' ||
 #endif
 	  *cp == '/')
Index: binutils/dllwrap.c
===================================================================
--- binutils/dllwrap.c	(Revision 5)
+++ binutils/dllwrap.c	(Arbeitskopie)
@@ -256,7 +256,7 @@
 	dash = cp;
 
       if (
-#if defined(__DJGPP__) || defined (__CYGWIN__) || defined(__WIN32__)
+#if defined(__DJGPP__) || defined (__CYGWIN__) || defined(__WIN32__) || defined(__MINT__)
 	  *cp == ':' || *cp == '\\' ||
 #endif
 	  *cp == '/')
Index: binutils/resrc.c
===================================================================
--- binutils/resrc.c	(Revision 5)
+++ binutils/resrc.c	(Arbeitskopie)
@@ -391,7 +391,7 @@
     *space = 0;
 
   if (
-#if defined (__DJGPP__) || defined (__CYGWIN__) || defined (_WIN32)
+#if defined (__DJGPP__) || defined (__CYGWIN__) || defined (_WIN32) || defined (__MINT__)
       strchr (cmd, '\\') ||
 #endif
       strchr (cmd, '/'))
@@ -509,7 +509,7 @@
 	  if (*cp == '-')
 	    dash = cp;
 	  if (
-#if defined (__DJGPP__) || defined (__CYGWIN__) || defined(_WIN32)
+#if defined (__DJGPP__) || defined (__CYGWIN__) || defined(_WIN32) || defined (__MINT__)
 	      *cp == ':' || *cp == '\\' ||
 #endif
 	      *cp == '/')
Index: binutils/testsuite/binutils-all/copy-2.d
===================================================================
--- binutils/testsuite/binutils-all/copy-2.d	(Revision 5)
+++ binutils/testsuite/binutils-all/copy-2.d	(Arbeitskopie)
@@ -3,7 +3,7 @@
 #objcopy: --set-section-flags foo=contents,alloc,load,code
 #name: copy with setting section flags 2
 #source: copytest.s
-#not-target: *-*-*aout *-*-aix* h8300-*-* hppa*-*-hpux* m68k-*-netbsd m68k-*-openbsd* mips-sony-bsd* mips-*-ultrix* mips-*-osf* mips-*-ecoff* mips-*-irix mips-*-irix[2-4] mips-*-riscos* ns32k-*-netbsd
+#not-target: *-*-*aout *-*-aix* h8300-*-* hppa*-*-hpux* m68k-*-netbsd m68k-*-openbsd* m68k-*-mint* mips-sony-bsd* mips-*-ultrix* mips-*-osf* mips-*-ecoff* mips-*-irix mips-*-irix[2-4] mips-*-riscos* ns32k-*-netbsd
 # Note - we use copytest.s and a section named "foo" rather
 # than .text because for some file formats (eg PE) the .text
 # section has a fixed set of flags and these cannot be changed.
Index: binutils/testsuite/binutils-all/copy-3.d
===================================================================
--- binutils/testsuite/binutils-all/copy-3.d	(Revision 5)
+++ binutils/testsuite/binutils-all/copy-3.d	(Arbeitskopie)
@@ -3,7 +3,7 @@
 #objcopy: --set-section-flags .text=alloc,data
 #name: copy with setting section flags 3
 #source: bintest.s
-#not-target: *-*-*aout *-*-*coff *-*-cygwin* *-*-darwin *-*-mingw* *-*-go32 *-*-*pe hppa*-*-hpux* m68k-*-netbsd m68k-*-openbsd* ns32k-*-netbsd powerpc-*-aix* rs6000-*-* rx-*-*
+#not-target: *-*-*aout *-*-*coff *-*-cygwin* *-*-darwin *-*-mingw* *-*-go32 *-*-*pe hppa*-*-hpux* m68k-*-mint* m68k-*-netbsd m68k-*-openbsd* ns32k-*-netbsd powerpc-*-aix* rs6000-*-* rx-*-*
 # The .text # section in PE/COFF has a fixed set of flags and these
 # cannot be changed.  We skip it for them.
 
Index: gas/config/te-mint.h
===================================================================
--- gas/config/te-mint.h	(nicht existent)
+++ gas/config/te-mint.h	(Arbeitskopie)
@@ -0,0 +1,31 @@
+/* Copyright 2008 Free Software Foundation, Inc.
+
+   This file is part of GAS, the GNU Assembler.
+
+   GAS is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as
+   published by the Free Software Foundation; either version 3,
+   or (at your option) any later version.
+
+   GAS is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
+   the GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with GAS; see the file COPYING.  If not, write to the Free
+   Software Foundation, 51 Franklin Street - Fifth Floor, Boston, MA
+   02110-1301, USA.  */
+
+#define TE_MINT
+
+#define LOCAL_LABELS_DOLLAR 1
+#define LOCAL_LABELS_FB 1
+#define REGISTER_PREFIX_OPTIONAL 1
+
+/* These define interfaces.  */
+#ifdef   OBJ_HEADER
+#include OBJ_HEADER
+#else
+#include "obj-format.h"
+#endif
Index: gas/configure.tgt
===================================================================
--- gas/configure.tgt	(Revision 5)
+++ gas/configure.tgt	(Arbeitskopie)
@@ -316,6 +316,7 @@
   m68k-*-linux-*)			fmt=elf em=linux ;;
   m68k-*-uclinux*)			fmt=elf em=uclinux ;;
   m68k-*-gnu*)				fmt=elf ;;
+  m68k-*-mint*)				fmt=elf em=mint bfd_gas=yes ;;
   m68k-*-netbsdelf*)			fmt=elf em=nbsd ;;
   m68k-*-netbsd*)			fmt=aout em=nbsd bfd_gas=yes ;;
   m68k-*-openbsd*)			fmt=aout em=nbsd bfd_gas=yes ;;
Index: gas/testsuite/gas/all/weakref1u.d
===================================================================
--- gas/testsuite/gas/all/weakref1u.d	(Revision 5)
+++ gas/testsuite/gas/all/weakref1u.d	(Arbeitskopie)
@@ -3,7 +3,7 @@
 #source: weakref1.s
 # aout turns undefined into *ABS* symbols.
 # see weakref1.d for comments on the other not-targets
-#not-target: *-*-*aout m68k-*-netbsd m68k-*-openbsd* ns32k-*-netbsd alpha*-*-osf* *-*-ecoff
+#not-target: *-*-*aout m68k-*-netbsd m68k-*-openbsd* m68k-*-mint* ns32k-*-netbsd alpha*-*-osf* *-*-ecoff
 
 # the rest of this file is generated with the following script:
 # # script begin
Index: gas/testsuite/gas/m68k/all.exp
===================================================================
--- gas/testsuite/gas/m68k/all.exp	(Revision 5)
+++ gas/testsuite/gas/m68k/all.exp	(Arbeitskopie)
@@ -93,7 +93,7 @@
     gas_test_error "p13050-1.s" "-march=isab" "moveb (d8,An,Xi),(d16,An) unsupported on isab"
     run_dump_test p13050-2
 
-    if { [istarget *-*-*aout] || [istarget *-*-netbsd] || [istarget *-*-openbsd*] } then {
+    if { [istarget *-*-*aout] || [istarget *-*-netbsd] || [istarget *-*-mint*] || [istarget *-*-openbsd*] } then {
       run_dump_test p3041
       run_dump_test p3041data
       run_dump_test p3041pcrel
Index: gas/testsuite/gas/m68k/br-isaa.d
===================================================================
--- gas/testsuite/gas/m68k/br-isaa.d	(Revision 5)
+++ gas/testsuite/gas/m68k/br-isaa.d	(Arbeitskopie)
@@ -1,7 +1,7 @@
 #name: br-isaa.d
 #objdump: -dr
 #as: -march=isaa -pcrel
-#not-target: *-*-*aout m68k-*-netbsd m68k-*-openbsd*
+#not-target: *-*-*aout m68k-*-netbsd m68k-*-openbsd* m68k-*-mint*
 
 .*:     file format .*
 
Index: gas/testsuite/gas/m68k/br-isab.d
===================================================================
--- gas/testsuite/gas/m68k/br-isab.d	(Revision 5)
+++ gas/testsuite/gas/m68k/br-isab.d	(Arbeitskopie)
@@ -1,7 +1,7 @@
 #name: br-isab.d
 #objdump: -dr
 #as: -march=isab -pcrel
-#not-target: *-*-*aout m68k-*-netbsd m68k-*-openbsd*
+#not-target: *-*-*aout m68k-*-netbsd m68k-*-openbsd* m68k-*-mint*
 
 .*:     file format .*
 
Index: gas/testsuite/gas/m68k/br-isac.d
===================================================================
--- gas/testsuite/gas/m68k/br-isac.d	(Revision 5)
+++ gas/testsuite/gas/m68k/br-isac.d	(Arbeitskopie)
@@ -1,7 +1,7 @@
 #name: br-isac.d
 #objdump: -dr
 #as: -march=isac -pcrel
-#not-target: *-*-*aout m68k-*-netbsd m68k-*-openbsd*
+#not-target: *-*-*aout m68k-*-netbsd m68k-*-openbsd* m68k-*-mint*
 
 .*:     file format .*
 
Index: include/filenames.h
===================================================================
--- include/filenames.h	(Revision 5)
+++ include/filenames.h	(Arbeitskopie)
@@ -32,7 +32,7 @@
 extern "C" {
 #endif
 
-#if defined(__MSDOS__) || defined(_WIN32) || defined(__OS2__) || defined (__CYGWIN__)
+#if defined(__MSDOS__) || defined(_WIN32) || defined(__OS2__) || defined (__CYGWIN__) || defined (__MINT__)
 #  ifndef HAVE_DOS_BASED_FILE_SYSTEM
 #    define HAVE_DOS_BASED_FILE_SYSTEM 1
 #  endif
Index: include/getopt.h
===================================================================
--- include/getopt.h	(Revision 5)
+++ include/getopt.h	(Arbeitskopie)
@@ -105,7 +105,7 @@
    to find the declaration so provide a fully prototyped one.  If it
    is 1, we found it so don't provide any declaration at all.  */
 #if !HAVE_DECL_GETOPT
-#if defined (__GNU_LIBRARY__) || defined (HAVE_DECL_GETOPT)
+#if defined (__GNU_LIBRARY__) || defined (__MINT__) || defined (HAVE_DECL_GETOPT)
 /* Many other libraries have conflicting prototypes for getopt, with
    differences in the consts, in unistd.h.  To avoid compilation
    errors, only prototype getopt for the GNU C library.  */
Index: ld/configure
===================================================================
--- ld/configure	(Revision 5)
+++ ld/configure	(Arbeitskopie)
@@ -16915,7 +16915,7 @@
 
 
 case "${host}" in
-*-*-msdos* | *-*-go32* | *-*-mingw32* | *-*-cygwin* | *-*-windows*)
+*-*-msdos* | *-*-go32* | *-*-mingw32* | *-*-cygwin* | *-*-windows* | *-*-mint*)
 
 $as_echo "#define USE_BINARY_FOPEN 1" >>confdefs.h
  ;;
Index: ld/configure.tgt
===================================================================
--- ld/configure.tgt	(Revision 5)
+++ ld/configure.tgt	(Arbeitskopie)
@@ -446,6 +446,7 @@
 			tdir_m68klinux=`echo ${targ_alias} | sed -e 's/linux/linuxaout/'` ;;
 m68k-*-uclinux*)	targ_emul=m68kelf ;;
 m68*-*-gnu*)		targ_emul=m68kelf ;;
+m68*-*-mint*)		targ_emul=m68kmint ;;
 m68*-*-netbsd*4k*)	targ_emul=m68k4knbsd
 			targ_extra_emuls="m68knbsd m68kelfnbsd" ;;
 m68*-*-netbsdelf*)	targ_emul=m68kelfnbsd
Index: ld/emulparams/m68kmint.sh
===================================================================
--- ld/emulparams/m68kmint.sh	(nicht existent)
+++ ld/emulparams/m68kmint.sh	(Arbeitskopie)
@@ -0,0 +1,20 @@
+SCRIPT_NAME=m68kmint
+OUTPUT_FORMAT="elf32-m68k"
+RELOCATEABLE_OUTPUT_FORMAT="elf32-m68k"
+#NO_REL_RELOCS=yes
+MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
+COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
+ARCH=m68k
+MACHINE=
+NOP=0x4e714e71
+TEMPLATE_NAME=elf32
+EXTRA_EM_FILE=m68kelf
+#GENERATE_SHLIB_SCRIPT=no
+#GENERATE_PIE_SCRIPT=no
+#NO_SMALL_DATA=yes
+#SCRIPT_NAME=m68kmint
+#OUTPUT_FORMAT="a.out-mintprg"
+#RELOCATEABLE_OUTPUT_FORMAT="a.out-zero-big"
+#TEXT_START_ADDR=0xe4
+#ARCH=m68k
+#EXTRA_EM_FILE=mint
Index: ld/Makefile.am
===================================================================
--- ld/Makefile.am	(Revision 5)
+++ ld/Makefile.am	(Arbeitskopie)
@@ -358,6 +358,7 @@
 	em68kelf.c \
 	em68kelfnbsd.c \
 	em68klinux.c \
+	em68kmint.c \
 	em68knbsd.c \
 	em68kpsos.c \
 	em88kbcs.c \
@@ -1555,6 +1556,10 @@
 em68klinux.c: $(srcdir)/emulparams/m68klinux.sh \
   $(srcdir)/emultempl/linux.em $(srcdir)/scripttempl/aout.sc ${GEN_DEPENDS}
 
+em68kmint.c: $(srcdir)/emulparams/m68kmint.sh \
+  $(srcdir)/emultempl/generic.em $(srcdir)/emultempl/m68kelf.em \
+  $(srcdir)/scripttempl/m68kmint.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} m68kmint "$(tdir_m68kmint)"
 em68knbsd.c:	$(srcdir)/emulparams/m68knbsd.sh \
   $(srcdir)/emultempl/generic.em $(srcdir)/scripttempl/aout.sc ${GEN_DEPENDS}
 
Index: ld/Makefile.in
===================================================================
--- ld/Makefile.in	(Revision 5)
+++ ld/Makefile.in	(Arbeitskopie)
@@ -726,6 +726,7 @@
 	em68kelf.c \
 	em68kelfnbsd.c \
 	em68klinux.c \
+	em68kmint.c \
 	em68knbsd.c \
 	em68kpsos.c \
 	em88kbcs.c \
@@ -1421,6 +1422,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/em68kelf.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/em68kelfnbsd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/em68klinux.Po@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/em68kmint.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/em68knbsd.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/em68kpsos.Po@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/em88kbcs.Po@am__quote@
@@ -3143,6 +3145,10 @@
 em68klinux.c: $(srcdir)/emulparams/m68klinux.sh \
   $(srcdir)/emultempl/linux.em $(srcdir)/scripttempl/aout.sc ${GEN_DEPENDS}
 
+em68kmint.c: $(srcdir)/emulparams/m68kmint.sh \
+  $(srcdir)/emultempl/generic.em $(srcdir)/emultempl/m68kelf.em \
+  $(srcdir)/scripttempl/m68kmint.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} m68kmint "$(tdir_m68kmint)"
 em68knbsd.c:	$(srcdir)/emulparams/m68knbsd.sh \
   $(srcdir)/emultempl/generic.em $(srcdir)/scripttempl/aout.sc ${GEN_DEPENDS}
 
Index: ld/scripttempl/m68kmint.sc
===================================================================
--- ld/scripttempl/m68kmint.sc	(nicht existent)
+++ ld/scripttempl/m68kmint.sc	(Arbeitskopie)
@@ -0,0 +1,84 @@
+cat <<EOF
+${RELOCATING+OUTPUT_FORMAT(${OUTPUT_FORMAT})}
+${RELOCATING-OUTPUT_FORMAT(${RELOCATEABLE_OUTPUT_FORMAT})}
+SECTIONS
+{
+  ${RELOCATING+".mint_prg_info (INFO) :
+  {
+    SHORT(0x601a);
+    LONG(.data_start - .text_start );
+    LONG(.bss_start - .data_start );
+    LONG(.prg_end - .bss_start );
+    LONG(0);
+    LONG(0x4d694e54);
+    LONG(0);
+    SHORT(0);
+  }"}
+  .text :
+  {
+    ${RELOCATING+.text_start = .;}
+
+    LONG(0x203a001a);                  /* g_jump_entry[0] */
+    LONG(0x4efb08fa);                  /* g_jump_entry[0] */
+    LONG(0x108);                       /* e_info */
+    LONG(.data_start - .entry );       /* e_text */
+    LONG(.bss_start - .data_start );   /* e_data */
+    LONG(.prg_end - .bss_start );      /* e_bss */
+    LONG(0);                           /* e_syms */
+    LONG(.entry);                      /* e_entry */
+    . = . + 0x10;
+    LONG(0);                           /* g_stkpos (filled from tostool) */
+    LONG(1);                           /* g_symbol_format */
+    . = . + 0xAC;
+    .entry = .;
+    KEEP(*(.text.entry.mint)) KEEP(*crt0*.o(.text)) KEEP(*startup.o(.text))
+    .startup_size = . - .entry;
+
+    *(.text.unlikely .text.*_unlikely)
+    *(.text.exit .text.exit.*)
+    *(.text.startup .text.startup.*)
+    *(.text.hot .text.hot.*)
+    *(.text .stub .text.* .gnu.linkonce.t.*)
+    *(.rodata .rodata.*)
+
+    /* .ctors */
+    ___CTOR_LIST__ = .;
+    LONG((___CTOR_END__ - ___CTOR_LIST__) / 4 - 2);
+    KEEP (*(.ctors))
+    KEEP (*(SORT(.ctors.*)))
+    LONG(0);
+    ___CTOR_END__ = .;
+
+    /* .dtors */
+    ___DTOR_LIST__ = .;
+    LONG((___DTOR_END__ - ___DTOR_LIST__) / 4 - 2);
+    KEEP (*(.dtors))
+    KEEP (*(SORT(.dtors.*)))
+    LONG(0);
+    ___DTOR_END__ = .;
+
+  }
+  ${RELOCATING+_etext = .;}
+  ${RELOCATING+__etext = .;}
+
+  .data . :
+  {
+    ${RELOCATING+.data_start = .;}
+    *(.data) *(.data.*) *(.gnu.linkonce.d.*)
+    ${RELOCATING+___EH_FRAME_BEGIN__ = . ;}
+    ${RELOCATING+*(.eh_fram*)}
+    ${RELOCATING+___EH_FRAME_END__ = . ;}
+  }
+  ${RELOCATING+_edata = .;}
+  ${RELOCATING+__edata = .;}
+
+  .bss . :
+  {
+    ${RELOCATING+.bss_start = .;}
+    *(.bss) *(COMMON) *(.bss.*) *(.gnu.linkonce.b.*)
+  }
+  ${RELOCATING+_end = .;}
+  ${RELOCATING+__end = .;}
+  ${RELOCATING+.prg_end = .;}
+}
+EOF
