Index: CONFIGVARS
===================================================================
--- CONFIGVARS	(Revision 1252)
+++ CONFIGVARS	(Arbeitskopie)
@@ -9,6 +9,7 @@
 #
 WITH_020_LIB = yes
 WITH_V4E_LIB = yes
+WITH_FASTCALL_LIB = yes
 
 #
 # target and host compiler
@@ -73,7 +74,7 @@
 #
 # optimizations
 # 
-OPTS = -O2 -fomit-frame-pointer
+OPTS = -O2 -fomit-frame-pointer $(EXTRA_OPTS)
 
 
 #
@@ -98,3 +99,28 @@
 # model type
 #
 MODEL = 
+
+# Multilib
+
+MULTILIBDIRS :=
+MULTILIBFLAGS =
+
+ifneq (,$(strip $(shell $(CC) -print-multi-lib | sed -e 's/\.;//')))
+	# gcc supports multilibs
+	# get multi-lib-dirs
+	MULTILIBDIRS := $(shell $(CC) -print-multi-lib | sed -e "s:;.*::")
+	ifneq ($(WITH_FASTCALL_LIB),yes)
+		# filter out fastcall
+		MULTILIBDIRS := $(shell echo $(MULTILIBDIRS) | sed -e 's/\S*mfastcall\S*/ /g')
+	endif
+	ifneq ($(WITH_V4E_LIB),yes)
+		# filter-out coldfire libs
+		MULTILIBDIRS := $(shell echo $(MULTILIBDIRS) | sed -e 's/\S*m5475\S*/ /g')
+	endif
+	ifneq ($(WITH_020_LIB),yes)
+		# filter-out 68020 libs
+		MULTILIBDIRS := $(shell echo $(MULTILIBDIRS) | sed -e 's/\S*m68020\S*/ /g')
+	endif
+	MULTILIBFLAGS = $(shell $(CC) -print-multi-lib | grep '^$(1);' | sed -e 's/^.*;//' -e 's/@/ -/g')
+
+endif
Index: DEPENDENCIES
===================================================================
--- DEPENDENCIES	(Revision 1252)
+++ DEPENDENCIES	(Arbeitskopie)
@@ -4,6 +4,7 @@
 $(CC) $(cflags) $(CFLAGS) $(CFLAGS-$(<F)) $(CFLAGS-$(DIRSTEM))) \
 $(DEFS-$(<F)) $(DEFS-$(DIRSTEM)))
 
+ifeq ($(MULTILIBDIRS),)
 #
 # compile rule
 #
@@ -21,6 +22,35 @@
 #
 DEP_FILES = $(addprefix ./.deps/, $(patsubst %.o,%.P,$(OBJS)))
 
+else
+
+DEP_FILES :=
+
+define CC_MULTI_TEMPLATE
+
+#
+# compile rule
+#
+
+.mlibs/$(1)/%.o : %.c
+	$$(AM_V_CC)$$(COMPILE)  -Wp,-MD,$$(@:.o=.d) -c $$< -o $$@
+.mlibs/$(1)/%.o : %.S
+	$$(AM_V_CPPAS)$$(COMPILE) -Wp,-MD,.$$(@:.o=.d) -c $$< -o $$@
+.mlibs/$(1)/%.o : %.S
+	$$(AM_V_AS)$$(COMPILE) -Wp,-MD,$$(@:.o=.d) -c $$< -o $$@
+
+#
+# dependencies
+#
+DEP_FILES += $(addprefix ./.mlibs/$(1)/, $(patsubst %.o,%.d,$(OBJS)))
+
+
+endef
+$(foreach DIR,$(MULTILIBDIRS),$(eval $(call CC_MULTI_TEMPLATE,$(DIR))))
+
+endif
+
+
 # These are automatically remade, no need for make to worry about
 # them.
 .PHONY : $(DEP_FILES)
Index: Makefile
===================================================================
--- Makefile	(Revision 1252)
+++ Makefile	(Arbeitskopie)
@@ -33,6 +33,9 @@
 GENFILES += mt_gem.h Doxyfile gemlib.spec
 GENFILES += html
 
+ifeq ($(MULTILIBDIRS),)
+	# gcc don't supports multilibs --> old stuff
+
 ALL_LIBS = normal mshort
 
 ifeq ($(WITH_020_LIB),yes)
@@ -43,6 +46,13 @@
 ALL_LIBS += m5475 m5475mshort
 endif
 
+else
+	ALL_LIBS = $(patsubst %,.mlibs/%/libgem.a, $(MULTILIBDIRS))
+endif
+
+
+
+
 libs: $(ALL_LIBS)
 
 normal:
@@ -131,6 +141,18 @@
 	make -f vbcc.mak;
 	make -f zip.mak
 
+define MULTI_TEMPLATE
+.mlibs/$(1) :
+	$$(AM_V_at)$$(MKDIR) -p $$@
+.mlibs/$(1)/libgem.a : GENERAL += $(call MULTILIBFLAGS,$(1))
+.mlibs/$(1)/libgem.a : mt_gem.h .mlibs/$(1) $(addprefix .mlibs/$(1)/, $(COBJS:.c=.o) $(SOBJS:.S=.o))
+	$$(AM_V_at)$$(RM) $$@
+	$$(AM_V_AR)$$(AR) rc $$@ $(addprefix .mlibs/$(1)/, $(COBJS:.c=.o) $(SOBJS:.S=.o))
+	$(AM_V_RANLIB)$(RANLIB) $$@
+
+endef
+$(foreach DIR,$(MULTILIBDIRS),$(eval $(call MULTI_TEMPLATE,$(DIR))))
+
 libgem$(TARGET).a: objs
 	$(AM_V_at)$(RM) $@
 	$(AM_V_AR)$(AR) rc $@ \
@@ -176,6 +198,8 @@
 #	$(MOVEIFCHANGE) $(srcdir)/mt_gem.tmp $(srcdir)/mt_gem.h
 
 install:
+ifeq ($(MULTILIBDIRS),)
+	# gcc don't supports multilibs --> old stuff
 	install -m 755 -d $(PREFIX)/lib
 	install -m 644 libgem.a $(PREFIX)/lib/libgem.a
 	install -m 755 -d $(PREFIX)/lib/mshort
@@ -193,6 +217,12 @@
 	install -m 755 -d $(PREFIX)/lib/m5475/mshort
 	install -m 644 libgemv4e16.a $(PREFIX)/lib/m5475/mshort/libgem.a
 endif
+else
+	for lib in $(MULTILIBDIRS); do \
+		install -m 755 -d $(PREFIX)/lib/$$lib; \
+		install -m 644 ./.mlibs/$$lib/libgem.a $(PREFIX)/lib/$$lib/libgem.a; \
+	done
+endif
 	install -m 755 -d $(PREFIX)/include
 	install -m 644 gem.h $(PREFIX)/include
 	install -m 644 gemx.h $(PREFIX)/include
@@ -200,6 +230,7 @@
 	install -m 644 mt_gemx.h $(PREFIX)/include
 
 uninstall:
+ifeq ($(MULTILIBDIRS),)
 	rm -f $(PREFIX)/lib/libgem.a
 	rm -f $(PREFIX)/lib/libgem16.a
 	rm -f $(PREFIX)/lib/mshort/libgem.a
@@ -211,7 +242,16 @@
 	rm -f $(PREFIX)/lib/m5475/libgem.a
 	rm -f $(PREFIX)/lib/m5475/mshort/libgem.a
 endif
+else
+	for lib in $(MULTILIBDIRS); do \
+		rm -f $(PREFIX)/lib/$$lib; \
+	done
+endif
 	rm -f $(PREFIX)/include/gem.h
 	rm -f $(PREFIX)/include/gemx.h
 	rm -f $(PREFIX)/include/mt_gem.h
 	rm -f $(PREFIX)/include/mt_gemx.h
+
+ifneq ($(MULTILIBDIRS),)
+include $(top_srcdir)/DEPENDENCIES
+endif
